### DISCUSSION TITLE
Create a new kernel with time derivative of the gradient of coupled variables, such as strain rate.

### DISCUSSION THREAD

**Discussion by Yaode-Yin** [idaholab/moose#29029] [Q&A Modules: General]
Hello MOOSE expert,
    I'm try to couple solid mechanics and the heat conduction. And some coupled terms should be added in the governing equation of heat transfer. 
    In the new heat equation, I have new terms related to the strain rate. So I think I need a new kernel for this term. So how can I create a new kernel with time derivative of the gradient of coupled variables, such as strain rate or the rate of order parameter?

**Comment by GiudGiud** [comment]
Hello

Have you followed the tutorial ? There is an example on how to create a new kernel.

The derivative of the gradient of the kernel can be obtained using the Coupleable interface
https://mooseframework.inl.gov/source/interfaces/Coupleable.html
but only for variables. An order parameter is a variable. The strain might be a material property?

**Reply by Yaode-Yin** [reply]
Hi, GiudGiud. Thanks for your reply. I have followed the tutorial of creating a Diffusion kernel. But I still cannot solve my problem. 

I'm try to solve a three-field coupled problem, and there are three nonlinear variables: displacements u =[ux uy uz], temperature T and order parameter c. 

In the scenario, the displacements and order parameter has added additional terms in heat conduction equation. For the order parameter, it has added the following term in heat equation:  c0 * T * grad(c)* dgrad(c)dt , where c0 is a material property.  And the weak form of this term should be c0 * T * grad(c) * dgrad(c)dt * test_function. 
And in MOOSE kernel file, the residual should be `_c0[_qp] * _phi[_qp] *_u[_qp] *_grad_v[_qp]*_grad_v_dot[_qp]*_test[_qp]`. I'm not sure whether the expression is right or not.

**Reply by GiudGiud** [reply]
```
c0 * T * grad(c) * dgrad(c)dt * test_function.
```
should translate to
```
_c0[_qp] *_u[_qp] *_grad_v[_qp]*_grad_v_dot[_qp]*_test[_qp]
```
if:
- T is the variable so _u is actually T
- grad_c has been coupled in using coupledGradient("c") to set grad_v
- same for dot

I recommend you upload your code to a git repository so you can share links here rather than paste extracts

**Reply by Yaode-Yin** [reply]
Thank you very much for your reply. My code has not finished yet. I'll upload it latter. 
I still have question on how to add a strain rate terms.   In my problem, the displacements also introduce additional term in heat equation, like:  

` c1 * sigma : depsilondt`
   
where c1 is a material property, sigma is a second-order tensor, epsilon is the small deformation strain, depsilondt is its rate.
I have no idea how to implement the term like this in MOOSE kernel system.

**Reply by GiudGiud** [reply]
c1 can be retrieved using getMaterialProperty, see any "Mat"Something kernel for an example
sigma can be a RealTensorValue. Is it going to be a material property or a constant?
epsilon, are you defining that one yourself or are you using the strain from the solid mechanics module?

**Reply by Yaode-Yin** [reply]
I think sigma should be a material property, because it is calculated with all the three field variables. Epsilon is the small deformation strain from the solid mechanics module.

**Reply by GiudGiud** [reply]
You can define tensors as material properties. So I think you can follow existing tensor material property examples to make it for now

**Reply by Yaode-Yin** [reply]
So do you mean I can create a kernel like c1* c2* c3 * test_function, and treat sigma and depsilondt as material properties? If so, do you have any example material property containing strain rate?

**Reply by GiudGiud** [reply]
I d look in the solid mechanics module for examples

**Reply by Yaode-Yin** [reply]
Hi, GiudGiud. I'm very new to crate moose kernel and struggling to implement the `CoupledGradientTimeDerivative`. It should be in the MOOSE framework, and computes the time derivative of the gradient of a coupled variable. I tried to implement this kernel based on the existing kernel `CoupledTimeDerivative`, but failed.  My code can be seen from https://github.com/Yaode-Yin/MOOSE_10_11_2024

**Reply by GiudGiud** [reply]
How did it fail? Is it not compiling or are you getting an error at runtime?

**Reply by Yaode-Yin** [reply]
It failed in `make` process, so I think it is not compiling.  The time derivative of the gradient of a coupled variable should be a vector as well as its Jacobian, but I dont know how to implement that in MOOSE.

**Reply by GiudGiud** [reply]
try 
```
const VariableGradient & 
```
instead of
```
const VectorMooseVariable & 
```
for the return type

**Reply by Yaode-Yin** [reply]
It still has some errors.

`In file included from /home/Desktop/moose/framework/build/unity_src/kernels_Unity.C:34:
/home/Desktop/moose/framework/src/kernels/CoupledGradTimeDerivative.C: In member function 'virtual libMesh::Real CoupledGradTimeDerivative::computeQpResidual()':
/home/Desktop/moose/framework/src/kernels/CoupledGradTimeDerivative.C:32:25: error: cannot convert 'libMesh::boostcopy::enable_if_c&lt;true, libMesh::TypeVector&lt;double&gt; &gt;::type' {aka 'libMesh::TypeVector&lt;double&gt;'} to 'libMesh::Real' {aka 'double'} in return
   32 |   return _test[_i][_qp] * _grad_v_dot[_qp];
make: *** [/home/Desktop/moose/framework/build.mk:151: /home/Desktop/moose/framework/build/unity_src/kernels_Unity.x86_64-conda-linux-gnu.opt.lo] Error 1`

I think that's because the residual should also be a vector.

**Reply by GiudGiud** [reply]
change ArrayKernel to Kernel in the constructor
also
```
return _test[_i][_qp] * _grad_v_dot[_qp]
```
returns a vector. You need to do something to get it down to a single value. Either use a dot product or take a component. 
Your weak form must have a mistake if the output of it comes to a vector instead of a single value

**Reply by Yaode-Yin** [reply]
I changed my strategy to implement the weak form, `_F[_qp] *_u[_qp] *_grad_v[_qp]*_grad_v_dot[_qp]*_test[_qp]`, in MOOSE, where F is a derivative material property.  I no longer try to write a generic kernel for the time derivative of the gradient of a coupled variable. Instead just implement my weak form.

Now my codes can be found here: https://github.com/Yaode-Yin/MOOSE_10_11_2024/blob/main/CoupledGradTimeDerivativeKernel.C
https://github.com/Yaode-Yin/MOOSE_10_11_2024/blob/main/CoupledGradTimeDerivativeKernel.h

But in the `make` process, I still meet some warnings: 
```
/home/Desktop/moose/modules/phase_field/build/header_symlinks/CoupledGradTimeDerivativeKernel.h: In constructor 'CoupledGradTimeDerivativeKernel::CoupledGradTimeDerivativeKernel(const InputParameters&amp;)':
/home/Desktop/moose/modules/phase_field/build/header_symlinks/CoupledGradTimeDerivativeKernel.h:40:34: warning: 'CoupledGradTimeDerivativeKernel::_dFdu' will be initialized after [-Wreorder]
   40 |   const MaterialProperty&lt;Real&gt; &amp; _dFdu;
      |                                  ^~~~~
/home/Desktop/moose/modules/phase_field/build/header_symlinks/CoupledGradTimeDerivativeKernel.h:33:28: warning:   'const VariableGradient&amp; CoupledGradTimeDerivativeKernel::_grad_v_dot' [-Wreorder]
   33 |   const VariableGradient &amp; _grad_v_dot;
      |                            ^~~~~~~~~~~
In file included from /home/Desktop/moose/modules/phase_field/build/unity_src/kernels_Unity.C:56:
/home/Desktop/moose/modules/phase_field/src/kernels/CoupledGradTimeDerivativeKernel.C:27:1: warning:   when initialized here [-Wreorder]
   27 | CoupledGradTimeDerivativeKernel::CoupledGradTimeDerivativeKernel(
```

I dont understand why these warnings are declared.  And how can I fix the code?

**Reply by GiudGiud** [reply]
in the constructor initialization list, order the attributes in the order as you have them in the header

**Reply by Yaode-Yin** [reply]
I got it, thanks a lot. I may be close to implement the kernel codes.

**Reply by Yaode-Yin** [reply]
Hi, GiudGiud. Do you know how to calculate the strain rate as a material property? I look for the material examples in solid mechanics module, but cannot find out related codes.

**Reply by GiudGiud** [reply]
There are a few examples in the solid mechanics materials. For example:
```
materials/ComputeIncrementalStrain.C:      _strain_rate[_qp] = _strain_increment[_qp] / _dt;
```
